src/Datadog.Trace/Agent/AgentWriterBuffer.cs

```diff
@@ -1,45 +1,52 @@
 using System;
-using System.Collections.Generic;
 
 namespace Datadog.Trace.Agent
 {
     internal class AgentWriterBuffer<T>
     {
-        private readonly object _lock = new object();
-        private readonly int _maxSize;
         private readonly Random _random = new Random();
-        private List<T> _items;
+        private readonly T[] _items;
+
+        private int _count;
 
         public AgentWriterBuffer(int maxSize)
         {
-            _maxSize = maxSize;
-            _items = new List<T>();
+            _items = new T[maxSize];
         }
 
         public bool Push(T item)
         {
-            lock (_lock)
+            lock (_items)
```

Comment:
`_items` is private, so although it is safe to apply the `lock` directly on it, the best practice for applying locks is to have a separate private `object` to lock, rather than use the data item you are guarding. The reason it makes sense is that, if you get in the habit of using the data item to apply a lock on, you might at some other time apply a lock to a publically-accessible item. Suggest using `private readonly object _itemsLock = new object();` earlier in this code.